runtime/compiler/spotbugs.vim

```diff
@@ -0,0 +1,49 @@
+" Vim compiler file
+" Compiler:     Spotbugs (Java static checker; needs javac compiled classes)
+" Maintainer:   @Konfekt
+" Last Change:  2024 nov 6
+
+if exists("current_compiler") | finish | endif
+let current_compiler = "spotbugs"
+
+let s:cpo_save = &cpo
+set cpo&vim
+
+" Unfortunately Spotbugs does not output absolute paths, so you need to
+" pass the directory of the files being checked as `-sourcepath` parameter.
+" The regex, auxpath and glob try to include all dependent classes of the
+" current buffer. See https://github.com/spotbugs/spotbugs/issues/856
+
+let s:slash = exists('+shellslash') && !&shellslash ? '\' : '/'
+let s:regex = '\v%(;|^)\s*%(public |protected |private )?\s*%(abstract |final )?\s*class\s+(\w+)%(\<.*\> )?\s+%(extends\s+\w+ |implements\s+\w+%(\s*,\s*\w+)* )?\s*%(permits\s+\w+%(\s*,\s*\w+)* )?\s*\{'
+
+function! s:GetClassFiles() abort
```

Comment:
@zzzyxwvut Would you mind adapting your patch to replace this function, leveraging syntax highlighting if enabled `if exists("g:syntax_on")`, and falling back to this otherwise `else`?
I did not understand from your reply if 

> So this snippet achieves matching class where it's used to define a class without any need to look for an opening brace?

Reply:
To only collect compilable types from the current buffer,  
we need to deal with comments and strings because it is not  
uncommon to include ‘type declarations’ in code snippets for  
Javadoc comments; it is also not uncommon to generate new  
source files, with the aid of annotation processors, from  
string templates (think [`String.format()`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/String.html#format(java.lang.String,java.lang.Object...))) that contain ‘type   
declarations’.  Thus, there can be three distinct contexts  
for ‘type declarations’ in some Java source files.

Now, for `class`es in the context of type declarations proper,  
we need to disambiguate class literals, e.g. `Object.class`,  
from `class` declarations; but such literals do not support  
block statements of `{}`.  Admittedly, there can be `record`  
[variables _or_ methods](https://github.com/vim/vim/blob/6065755a39d838aaba03808f77d58c34b3b63e91/runtime/syntax/testdir/input/java_contextual_keywords.java#L26-L27), but the type name in the regexp  
should reject most misidentified _and_ valid uses (but not  
`record instanceof Foo`, for which filename globbing would  
come to the rescue).

All in all, looking for `{` should only matter _inside of_  
comments and strings, where keywords are not enforceable.

Reply:
Thank you! I applied your patch, which works fine so far.

> To only collect compilable types from the current buffer, we need to deal with comments and strings because it is not uncommon to include ‘type declarations’ in code snippets for Javadoc comments; it is also not uncommon to generate new source files, with the aid of annotation processors, from string templates (think [`String.format()`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/String.html#format(java.lang.String,java.lang.Object...))) that contain ‘type declarations’. Thus, there can be three distinct contexts for ‘type declarations’ in some Java source files.
> 
> Now, for `class`es in the context of type declarations proper, we need to disambiguate class literals, e.g. `Object.class`, from `class` declarations; but such literals do not support block statements of `{}`. Admittedly, there can be `record` [variables _or_ methods](https://github.com/vim/vim/blob/6065755a39d838aaba03808f77d58c34b3b63e91/runtime/syntax/testdir/input/java_contextual_keywords.java#L26-L27), but the type name in the regexp should reject most misidentified _and_ valid uses (but not `record instanceof Foo`, for which filename globbing would come to the rescue).
> 
> All in all, looking for `{` should only matter _inside of_ comments and strings, where keywords are not enforceable.

I guess you are saying that by exclusion the only valid appearance of `class` are those requiring a brace, but thanks to knowing this there's no longer any need to explicitly check for it.